File: P1/Sesion1/FactoriaAbstracta_Prototipos_RUBY/lib/factoria_carrera_y_bicicleta.rb
module FactoriaAbstracta_Prototipos_RUBY

  class FactoriaCarreraYBicicleta

    attr_accessor :_prototipoCarreraMontana, 
                  :_prototipoCarreraCarretera, 
                  :_prototipoBicicletaCarretera,
                  :_prototipoBicicletaMontana

    def initialize
      @_prototipoCarreraMontaña = CarreraMontana.new
      @_prototipoCarreraCarretera   = CarreraCarretera.new
      @_prototipoBicicletaCarretera = BicicletaCarretera.new
      @_prototipoBicicletaMontaña   = BicicletaMontana.new
    end


    def crearCarrera(n, tipo)
      if(tipo == Tipo::CARRETERA)
          return _prototipoCarreraCarretera.clone();
      else
          return _prototipoCarreraMontana.clone();
      end
    end

    def crearBicicleta(tipo)
      if(tipo == Tipo::CARRETERA)
          return _prototipoBicicletaCarretera.clone();
      else
          return _prototipoBicicletaMontana.clone();
      end
    end

  end

end

File: P1/Sesion1/FactoriaAbstracta_Prototipos/src/prototipo/FactoriaCarreraYBicicleta.java
package prototipo;

public class FactoriaCarreraYBicicleta {

    private static CarreraMontaña      _prototipoCarreraMontaña;
    private static CarreraCarretera    _prototipoCarreraCarretera;
    private static BicicletaCarretera  _prototipoBicicletaCarretera;
    private static BicicletaMontaña    _prototipoBicicletaMontaña;

    
    /**
     * 
     */
    FactoriaCarreraYBicicleta() {
        _prototipoCarreraMontaña     = new CarreraMontaña(null);
        _prototipoCarreraCarretera   = new CarreraCarretera(null);
        _prototipoBicicletaCarretera = new BicicletaCarretera();
        _prototipoBicicletaMontaña   = new BicicletaMontaña();
    }
    
    
    /**
     * 
     * @param n
     * @param tipo
     */
    Carrera crearCarrera(int n, Tipo tipo) {
        if(tipo == Tipo.CARRETERA)
            return _prototipoCarreraCarretera.clone();
        else
            return _prototipoCarreraMontaña.clone();
    }

    
    /**
     * 
     * @param tipo
     */
    Bicicleta crearBicicleta(Tipo tipo) {
        // TODO - implement FactoriaCarreraYBicicleta.crearBicicleta
        throw new UnsupportedOperationException();
    }

}